mbf_abstract_nav/src/abstract_controller_execution.cpp

```diff
@@ -254,18 +254,24 @@ bool AbstractControllerExecution::reachedGoalCheck()
 
 bool AbstractControllerExecution::cancel()
 {
-  // request the controller to cancel; it returns false if cancel is not implemented or rejected by the plugin
-  if (!controller_->cancel())
+  // Request the controller to cancel; returns true if he takes care of stopping, returning CANCELED when done,
+  // allowing for smooth, controlled stops.
+  // If false, meaning cancel is not implemented, or that the controller defers handling it, MBF will take care
+  if (controller_->cancel())
   {
-    ROS_WARN_STREAM("Cancel controlling failed. Wait until the current control cycle finished!");
+    ROS_INFO("Controller will take care of stopping");
   }
-  // then wait for the control cycle to stop (should happen immediately if the controller cancel returned true)
-  cancel_ = true;
-  if (waitForStateUpdate(boost::chrono::milliseconds(500)) == boost::cv_status::timeout)
+  else
   {
-    // this situation should never happen; if it does, the action server will be unready for goals immediately sent
-    ROS_WARN_STREAM("Timeout while waiting for control cycle to stop; immediately sent goals can get stuck");
-    return false;
+    ROS_WARN("Controller defers handling cancel; force it and wait until the current control cycle finished");
+    cancel_ = true;
```

Comment:
should we set it also for the other case?

Reply:
in the other case we set it [here](https://github.com/magazino/move_base_flex/pull/274/files#diff-bbabf6c9cc39ada410c0f58e40f9cf9c80a104f403ee9196f6a1519fe972a2dfR394), once we are done with the cancel

Reply:
I was thinking just to tolerate if the user implements the interface wrong (say, returning SUCCESS instead of CANCEL)

Reply:
any opinion here?

Reply:
but then on [this line](https://github.com/magazino/move_base_flex/pull/274/files#diff-bbabf6c9cc39ada410c0f58e40f9cf9c80a104f403ee9196f6a1519fe972a2dfL295) we will butch the smooth stopping, isn't it?

Reply:
ah right